### ISSUE TITLE
rdata! macro does not work with generic types (lifetime specifier)

### ISSUE THREAD

**Issue by thorlucas** [fleabitdev/glsp#1]
I need to have a generic type in order to specify a lifetime. This is what the declaration looks like:

``` rust
struct DAC<'a> {
    devices: Vec<Box<dyn Device + 'a>>,
}
```

Unfortunately, wrapping this in a `rdata!` macro throws an error:

``` rust
rdata! {
    struct DAC<'a> {
        devices: Vec<Box<dyn Device + 'a>>,
    }
}
```

```
[rustc] [E] no rules expected the token `<`
no rules expected this token in macro call
```

I'm unsure of how to use this as a global variable in glsp if this is the case.

The goal here is to have this function:

``` rust
pub fn add_device<D>(&mut self, device: D)
where
    D: Device + 'a,
{
    self.devices.push(Box::new(device));
}
```

Unfortunately it seems like I also cannot just define a closure that captures the `dac`, because I can't make a generic closure.

In the meanwhile, I'm open to any alternatives that accomplish the following: after `main.glsp` is run, I need a list of structs that implement the `Device` trait. The `main.glsp` builds up these devices and adds them to the list.

**Comment by thorlucas** [comment]
Additionally, I can't use just `Arr` because of this:

``` rust
rdata! {
    struct DAC {
        devices: Arr,
    }
}

impl DAC {
    pub fn new() -> Self {
        DAC { devices: arr![] }
    }

    pub fn add_device(&mut self, device: Val) {
        self.devices.push(device);
    }
}
```

This doesn't work because `arr![]` returns `Root<Arr>`. So, let's try that:

``` rust
rdata! {
    struct DAC {
        devices: Root<Arr>,
    }
}
```

This doesn't work because `Root<Arr>` does not satisfy `GStore`.

**Comment by fleabitdev** [comment]
Thanks for getting in touch!

The `RStore` trait has a `: 'static` bound, because data promoted to the GameLisp heap needs to be able to live at least as long as the enclosing `Runtime` - any `rdata` could potentially be stored in a global variable, and then accessed by some remote part of the codebase using `glsp::global`.

I agree that I should support lifetime parameters and other type parameters syntactically, but it would currently only be possible to pass such a value to `glsp::rdata` when the lifetime parameter is `'static`. Allowing borrowed Rust data to be moved onto the GC heap would be a significant engineering challenge.

It seems as though the main roadblock you're running into here is that the data you can store on GameLisp's heap is quite restricted. It [can't store Roots](https://gamelisp.rs/reference/rust-data.html#internal-references), can't store non-`'static` data, can't store any custom structs which don't implement `RStore`.

To solve the problem as written, your Rust code might look like this:

    //in order to store `dyn Device` on the glsp heap, you need to assert that each Device type
    //would, itself, be valid to store on the heap, and then declare those types using rdata!
    trait Device: glsp::GStore {
        fn be_a_device(&self) { }
    }

    //Box<T> doesn't currently implement RStore, so you need a wrapper. (i could probably
    //change this, come to think of it)
    rdata! {
        struct DeviceWrapper(Box<dyn Device>);
    }

    rdata! {
        struct Devices {
            devices: Vec<Box<dyn Device>>
        }
    }

    impl Devices {
        //Rust data is stored on the glsp heap by value, not by reference. to take ownership
        //of a Device, we need to remove it from the heap using RData::take()
        fn add_device(&mut self, src: Root<RData>) -> GResult<()> {
            let wrapper: DeviceWrapper = src.take()?;
            self.devices.push(wrapper.0);
            Ok(())
        }
    }

You have several options for simplifying this:

- Manipulate your list of `Devices` as an untyped `Root<Arr>` on the Rust stack, since it's just an array of `rdata`. You can push or pop `Root<RData>` or `RRoot<DeviceWrapper>` from an `Arr`.
- If you need the list of `Devices` to be persistent and globally available, you could store them in a `Lib` rather than an `RData`.
- If so, you could refer to your devices indirectly using `RRoot<DeviceWrapper>`, rather than taking exclusive ownership of them using `Box<dyn Device>`.
- If you don't need to support completely arbitrary `Device` types, you could consider replacing your `Device` trait with a simple enum.

Leaving this issue open with a TODO list for myself:
- Syntactically support generic parameters within `rdata!`
- ~~Support `rdata! { enum ... }`~~
- ~~Document `RStore`'s `'static` bound, and non-`Root` restriction, more thoroughly~~
- ~~Document the fact that `Arr` is manipulated as `Root<Arr>`, etc., more thoroughly~~
- Consider implementing `RStore` for standard library types like `Box` and `Vec`

**Comment by thorlucas** [comment]
Thanks! This project is really awesome — I fell in love with Lisp-like languages when I learned Berkeley Scheme in school.

The project I started working on a few days ago is a [PureData](https://puredata.info/)-like playground, using `cpal` for audio generation and `tui-rs` as the user interface. I knew I would need a scripting language eventually, and when I was this on Reddit I was very impressed and excited to use it.

Essentially, I'm trying to expose an scripting interface that looks something like this:

``` lisp
; adds a sin wave at 440.0 Hz and 0.0 degrees
(add-device
    (osc:sin:const 440.0 0.0))

; the knob is a logarithmic knob that goes from 20.0Hz to 20,000.0Hz
; the osc is a variable sin wave controlled by a knob
; the osc has a reference to the knob, or some sort of ID to it or something

; add-device returns an ID or a reference or something to the added device
(add-device
    (osc:sin:var
        (add-device 
            (knob 20.0 200000.0 'log))
        0.0))
```

Additionally I plan to add support for convolutional filters, wavetables, etc. 

I hope I can do something like this with glsp!

**Comment by fleabitdev** [comment]
Sounds interesting! And it does sound like a good fit for GameLisp. There aren't many other scripting languages which are safe to use for audio processing, because of the risk that a garbage collection might cause a gap in the audio stream.

In this case, yes, unfortunately you're running face-first into the fact that GameLisp can't store GC-to-GC pointers. It'll be fixed one day, but it's not on the roadmap just yet. Functions like `add-device` can return an `rdata` reference, but the actual `Device` struct will need to store an integer ID which looks up an `RRoot<Device>` from a library or from a global array.

Keep me updated if you run into any other problems!

**Comment by thorlucas** [comment]
The `lib!` approach seems to work a charm. Thank you!

**Comment by fleabitdev** [comment]
I've implemented some of my TODO list above, but I've decided to postpone the generic-types-as-`RData` feature, including standard library types like `Box<T: RStore>` and `Vec<T: RStore>`. There are a few major obstacles:

- I'd prefer not to pull in `syn` as a dependency, so `rdata!` is a `macro_rules`-based macro. Generic parameter lists, and `where` clauses, would be a nightmare to parse using the current `macro_rules` syntax. Likewise, emitting generic trait implementations from a `macro_rules` macro would be difficult.

- We currently print `RData` types, and look up their associated `RClass`, by calling a trait function with the signature `RStore::type_name() -> &'static str`. There's currently no reliable way to produce a `&'static str` which names a generic type. Returning a `String` instead could lead to performance problems.

- Storing Rust's collection types on the GameLisp heap might be counterintuitive for new users, because it probably won't be possible to index those types from GameLisp code. Likewise, if we support lifetime parameters syntactically, the `'static` bound on `RStore` might trip some people up.

I might revisit this if we ever decide to add `syn` as a default dependency.
